### ISSUE TITLE
Error on changing theme in app

### ISSUE THREAD

**Issue by muhammadmateen027** [kherel/animated_theme_switcher#36]
Facing this error while changing theme: 

```dart
[log] Null check operator used on a null value
      #0      ThemeSwitcherCircleClipper.getClip (package:animated_theme_switcher/src/clippers/theme_switcher_circle_clipper.dart:15:25)
      #1      ThemeSwitcherClipperBridge.getClip (package:animated_theme_switcher/src/clippers/theme_switcher_clipper_bridge.dart:14:21)
      #2      _RenderCustomClip._updateClip (package:flutter/src/rendering/proxy_box.dart:1375:25)
      #3      RenderClipPath.paint (package:flutter/src/rendering/proxy_box.dart:1689:7)
      #4      RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #5      PaintingContext.paintChild (package:flutter/src/rendering/object.dart:187:13)
      #6      RenderBoxContainerDefaultsMixin.defaultPaint (package:flutter/src/rendering/box.dart:2795:15)
      #7      RenderStack.paintStack (package:flutter/src/rendering/stack.dart:629:5)
      #8      RenderStack.paint (package:flutter/src/rendering/stack.dart:645:7)
      #9      RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #10     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:187:13)
      #11     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:140:15)
      #12     _RenderInkFeatures.paint (package:flutter/src/material/material.dart:551:11)
      #13     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #14     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:187:13)
      #15     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:140:15)
      #16     PaintingContext.pushLayer (package:flutter/src/rendering/object.dart:394:12)
      #17     RenderPhysicalModel.paint (package:flutter/src/rendering/proxy_box.dart:1922:15)
      #18     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #19     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:187:13)
      #20     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:140:15)
      #21     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #22     PaintingContext._repaintCompositedChild (package:flutter/src/rendering/object.dart:139:11)
      #23     PaintingContext.repaintCompositedChild (package:flutter/src/rendering/object.dart:100:5)
      #24     PipelineOwner.flushPaint (package:flutter/src/rendering/object.dart:975:29)
      #25     RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:464:19)
      #26     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:876:13)
      #27     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:328:5)
      #28     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1144:15)
      #29     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1082:9)
      #30     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:998:5)
      #31     _rootRun (dart:async/zone.dart:1354:13)
      #32     _CustomZone.run (dart:async/zone.dart:1258:19)
      #33     _CustomZone.runGuarded (dart:async/zone.dart:1162:7)
      #34     _invoke (dart:ui/hooks.dart:163:10)
      #35     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:259:5)
      #36     _drawFrame (dart:ui/hooks.dart:126:31)
[log] Null check operator used on a null value
      #0      ThemeSwitcherCircleClipper.getClip (package:animated_theme_switcher/src/clippers/theme_switcher_circle_clipper.dart:15:25)
      #1      ThemeSwitcherClipperBridge.getClip (package:animated_theme_switcher/src/clippers/theme_switcher_clipper_bridge.dart:14:21)
      #2      _RenderCustomClip._updateClip (package:flutter/src/rendering/proxy_box.dart:1375:25)
      #3      RenderClipPath.paint (package:flutter/src/rendering/proxy_box.dart:1689:7)
      #4      RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #5      PaintingContext.paintChild (package:flutter/src/rendering/object.dart:187:13)
      #6      RenderBoxContainerDefaultsMixin.defaultPaint (package:flutter/src/rendering/box.dart:2795:15)
      #7      RenderStack.paintStack (package:flutter/src/rendering/stack.dart:629:5)
      #8      RenderStack.paint (package:flutter/src/rendering/stack.dart:645:7)
      #9      RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #10     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:187:13)
      #11     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:140:15)
      #12     _RenderInkFeatures.paint (package:flutter/src/material/material.dart:551:11)
      #13     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #14     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:187:13)
      #15     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:140:15)
      #16     PaintingContext.pushLayer (package:flutter/src/rendering/object.dart:394:12)
      #17     RenderPhysicalModel.paint (package:flutter/src/rendering/proxy_box.dart:1922:15)
      #18     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #19     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:187:13)
      #20     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:140:15)
      #21     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #22     PaintingContext._repaintCompositedChild (package:flutter/src/rendering/object.dart:139:11)
      #23     PaintingContext.repaintCompositedChild (package:flutter/src/rendering/object.dart:100:5)
      #24     PipelineOwner.flushPaint (package:flutter/src/rendering/object.dart:975:29)
      #25     RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:464:19)
      #26     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:876:13)
      #27     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:328:5)
      #28     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1144:15)
      #29     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1082:9)
      #30     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:998:5)
      #31     _rootRun (dart:async/zone.dart:1354:13)
      #32     _CustomZone.run (dart:async/zone.dart:1258:19)
      #33     _CustomZone.runGuarded (dart:async/zone.dart:1162:7)
      #34     _invoke (dart:ui/hooks.dart:163:10)
      #35     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:259:5)
      #36     _drawFrame (dart:ui/hooks.dart:126:31)
[log] Null check operator used on a null value
      #0      ThemeSwitcherCircleClipper.getClip (package:animated_theme_switcher/src/clippers/theme_switcher_circle_clipper.dart:15:25)
      #1      ThemeSwitcherClipperBridge.getClip (package:animated_theme_switcher/src/clippers/theme_switcher_clipper_bridge.dart:14:21)
      #2      _RenderCustomClip._updateClip (package:flutter/src/rendering/proxy_box.dart:1375:25)
      #3      RenderClipPath.paint (package:flutter/src/rendering/proxy_box.dart:1689:7)
      #4      RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #5      PaintingContext.paintChild (package:flutter/src/rendering/object.dart:187:13)
      #6      RenderBoxContainerDefaultsMixin.defaultPaint (package:flutter/src/rendering/box.dart:2795:15)
      #7      RenderStack.paintStack (package:flutter/src/rendering/stack.dart:629:5)
      #8      RenderStack.paint (package:flutter/src/rendering/stack.dart:645:7)
      #9      RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #10     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:187:13)
      #11     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:140:15)
      #12     _RenderInkFeatures.paint (package:flutter/src/material/material.dart:551:11)
      #13     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #14     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:187:13)
      #15     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:140:15)
      #16     PaintingContext.pushLayer (package:flutter/src/rendering/object.dart:394:12)
      #17     RenderPhysicalModel.paint (package:flutter/src/rendering/proxy_box.dart:1922:15)
      #18     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #19     PaintingContext.paintChild (package:flutter/src/rendering/object.dart:187:13)
      #20     RenderProxyBoxMixin.paint (package:flutter/src/rendering/proxy_box.dart:140:15)
      #21     RenderObject._paintWithContext (package:flutter/src/rendering/object.dart:2317:7)
      #22     PaintingContext._repaintCompositedChild (package:flutter/src/rendering/object.dart:139:11)
      #23     PaintingContext.repaintCompositedChild (package:flutter/src/rendering/object.dart:100:5)
      #24     PipelineOwner.flushPaint (package:flutter/src/rendering/object.dart:975:29)
      #25     RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:464:19)
      #26     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:876:13)
      #27     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:328:5)
      #28     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1144:15)
      #29     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1082:9)
      #30     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:998:5)
      #31     _rootRun (dart:async/zone.dart:1354:13)
      #32     _CustomZone.run (dart:async/zone.dart:1258:19)
      #33     _CustomZone.runGuarded (dart:async/zone.dart:1162:7)
      #34     _invoke (dart:ui/hooks.dart:163:10)
      #35     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:259:5)
      #36     _drawFrame (dart:ui/hooks.dart:126:31)
```

Code for this page:

```dart
import 'dart:developer';

import 'package:animated_theme_switcher/animated_theme_switcher.dart';
import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';
import '../app/theme_config.dart';
import '../issues/issues.dart';

import 'components/components.dart';

class IssuesPage extends StatefulWidget {
  const IssuesPage({Key? key}) : super(key: key);

  @override
  _IssuesPageState createState() => _IssuesPageState();
}

class _IssuesPageState extends State<IssuesPage> {
  @override
  Widget build(BuildContext context) {
    return ThemeSwitchingArea(
      child: Scaffold(
        appBar: AppBar(
          title: const Text('Issues'),
          leading: _themeSwitcherButton(),
        ),
        body: const IssuesList(),
      ),
    );
  }

  Widget _themeSwitcherButton() {
    return ThemeSwitcher(
      clipper: const ThemeSwitcherCircleClipper(),
      builder: (context) {
        return IconButton(
          icon: const Icon(Icons.dark_mode),
          onPressed: () {
            ThemeSwitcher.of(context)!.changeTheme(
              theme: ThemeProvider.of(context)!.brightness == Brightness.light
                  ? darkTheme
                  : lightTheme,
            );
          },
        );
      },
    );
  }
}
```

**Flutter version:** 2.2.3

**Comment by videah** [comment]
From some poking around the issue seems to come from the use of `const` on a parent widget of `ThemeSwitchingArea`. Because of this `didUpdateWidget` doesn't get called on `ThemeSwitchingArea` so the offset for the clipper doesn't get set (resulting in a crash because it's null)

I don't know enough about the library to write up a fix yet but check where you're using `const` to fix it for now.

**Comment by Faddy07** [comment]
I also suffer from this issue. Is therer any solution for this?

**Comment by kherel** [comment]
Sorry guys, very busy right now.. I'll check it in 10 days. Feel free to make a pull request.

**Comment by skaldek** [comment]
Any updates?

**Comment by kherel** [comment]
I cannot reproduce the issue could you provide the full example

**Comment by skaldek** [comment]
@kherel Just add `const` to some widgets from the example. Below is the required part of example code. `const` is added in two places. 
```dart
void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    final isPlatformDark =
        WidgetsBinding.instance!.window.platformBrightness == Brightness.dark;
    final initTheme = isPlatformDark ? darkTheme : lightTheme;
    return ThemeProvider(
      initTheme: initTheme,
      builder: (_, myTheme) {
        return MaterialApp(
          title: 'Flutter Demo',
          theme: myTheme,
          home: const MyHomePage(),
        );
      },
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}
```

**Comment by kherel** [comment]
got it, thanks. on it..

**Comment by kherel** [comment]
The bug is fixed in version 2.0.3
